### DISCUSSION TITLE
cypress-image-snapshot

### DISCUSSION THREAD

**Discussion by SEVCAN94** [cypress-io/cypress#26683] [Questions and Help]
Hi,

I need to ask you a few questions. I want to write some tests which is going to compare screenshots. My main purpose is comparing screenshots of a page to be sure nothing changed after developments. 
I'm planning to use cypress-image-snapshot plugin but I also open for other suggestions if you have. While using cypress-image-snapshot plugin I've faced a few issues. Let me explain them.

1. Can we use cypress-image-snapshot plugin with latest Cypress versions? I'm asking this because couldn't run matchImageSnapshot test function while using Cypress 12.11.0 version. If we can use latest versions is there anything I should configure and how can I do?

2. I could run matchImageSnapshot test function when I was using Cypress 4.12.1 version but I cannot simulate the press of the Tab key. I have tried **cy.focused().type('{tab}')** , **cy.trigger('keydown', { keyCode: 9, which: 9 })** but it didn't work. How can I simulate the press of the Tab key while using matchImageSnapshot() ?

3.  My last question is I want to take the screenshot in the same moment hover over an element. How can I create this scenario?

**Comment by portyu9** [comment]
Hi @SEVCAN94 

There are three separate questions here.

### 1. `cypress-image-snapshot` with newer Cypress

The original `jaredpalmer/cypress-image-snapshot` package is the problem here, not Cypress's screenshot support. Development of that original plugin stopped and it was not updated properly for the Cypress 10+ plugin/configuration changes.

Today, use the actively maintained fork:

```bash
npm install --save-dev @simonsmith/cypress-image-snapshot
```

For a current Cypress project its setup is:

```ts
// cypress.config.ts
import { defineConfig } from 'cypress'
import {
  addMatchImageSnapshotPlugin,
} from '@simonsmith/cypress-image-snapshot/plugin'

export default defineConfig({
  e2e: {
    setupNodeEvents(on) {
      addMatchImageSnapshotPlugin(on)
    },
  },
})
```

and:

```ts
// cypress/support/e2e.ts
import {
  addMatchImageSnapshotCommand,
} from '@simonsmith/cypress-image-snapshot/command'

addMatchImageSnapshotCommand()
```

If you must remain on Cypress 12, use a release of that fork whose peer compatibility includes Cypress 12 rather than installing its newest release blindly.

Cypress itself provides `cy.screenshot()`, but Cypress does **not** perform image comparison natively.

### 2. Pressing Tab

This does not work:

```js
cy.focused().type('{tab}')
```

and manually triggering `keydown` does not reproduce the browser's native focus traversal either.

Current Cypress has a native keyboard command:

```js
cy.get('input').focus()

cy.press(Cypress.Keyboard.Keys.TAB)

cy.get('next-selector').should('have.focus')
```

`cy.press()` was added in Cypress 14.3.0. If you remain on Cypress 4.12.1, that API does not exist, so upgrade or use a compatible keyboard/tab plugin.

### 3. Screenshot while hovering

For JavaScript-driven hover behavior:

```js
cy.get('[data-cy="target"]').trigger('mouseover')

cy.screenshot('hover-state')
```

However, `.trigger('mouseover')` does **not** activate CSS `:hover`.

For a real Chromium hover, `cypress-real-events` provides `realHover()`:

```js
cy.get('[data-cy="target"]').realHover()

cy.screenshot('hover-state')
```

or run your image-snapshot command after `realHover()`.

So the right solution depends on whether the hover state comes from JavaScript events or CSS `:hover`.

There were no screenshots/attachments in the discussion that needed separate analysis.

Please mark this answer as accepted if it helped, thank you.
